packages/auth/src/providers/cognito/apis/signInWithRedirect.ts

```diff
@@ -0,0 +1,391 @@
+		return;
+	} finally {
+	}
+
+	await cacheCognitoTokens({
+		AccessToken: access_token,
+		IdToken: id_token,
+		RefreshToken: undefined,
+		TokenType: token_type,
+		ExpiresIn: expires_in,
+	});
+
+	// clear history
+
+	if (window && typeof window.history !== 'undefined') {
+		window.history.replaceState({}, null, redirectUri);
+	}
+
+	resolveInflight();
+
+	resolveInflight = () => {};
+}
+
+async function handleAuthResponse({
+	currentUrl,
+	userAgentValue,
+	clientId,
+	redirectUri,
+	responseType,
+	domain,
+}: {
+	currentUrl?: string;
+	userAgentValue?: string;
+	clientId: string;
+	redirectUri: string;
+	responseType: string;
+	domain: string;
+}) {
+	try {
+		const urlParams = new URL(currentUrl);
+		const error = urlParams.searchParams.get('error');
+		const error_description = urlParams.searchParams.get('error_description');
+
+		if (error) {
+			throw new AuthError(AuthErrorTypes.OAuthSignInError, error_description);
+		}
+
+		if (responseType === 'code') {
+			return await handleCodeFlow({
+				currentUrl,
+				userAgentValue,
+				clientId,
+				redirectUri,
+				domain,
+			});
+		} else {
+			return await handleImplicitFlow({
+				currentUrl,
+				redirectUri,
+			});
+		}
+	} catch (e) {
+		throw e;
+	}
+}
+
+async function validateStateFromURL(urlParams: URL): Promise<string> {
+	if (!urlParams) {
+		return;
+	}
+	const returnedState = urlParams.searchParams.get('state');
+
+	await validateState(returnedState);
+	return returnedState;
+}
+
+async function validateState(state: string) {
+	const savedState = await store.loadOAuthState();
+
+	// This is because savedState only exists if the flow was initiated by Amplify
+	if (savedState && savedState !== state) {
+		throw new AmplifyError({
+			name: '',
+			message: '',
+			recoverySuggestion: '',
+		});
+	}
+}
+
+function urlListener() {
+	// Listen configure to parse url
+	// TODO(v6): what happens if configure gets called multiple times during code exchange
+	Hub.listen('core', async capsule => {
+		if (capsule.payload.event === 'configure') {
+			const authConfig = AmplifyV6.getConfig().Auth;
+			store.setAuthConfig(authConfig);
+
+			// No OAuth inflight doesnt need to parse the url
+			if (!(await store.loadOAuthInFlight())) {
+				return;
+			}
+			try {
+				assertOAuthConfig(authConfig);
+			} catch (err) {
+				// TODO(v6): this should warn you have signInWithRedirect but is not configured
+				return;
+			}
+
+			try {
+				const url = window.location.href;
+
+				handleAuthResponse({
+					currentUrl: url,
+					clientId: authConfig.userPoolWebClientId,
+					domain: authConfig.oauth.domain,
+					redirectUri: authConfig.oauth.redirectSignIn,
+					responseType: authConfig.oauth.responseType,
+				});
+			} catch (err) {
+				// is ok if there is not OAuthConfig
+			}
+		}
+	});
+}
+
+urlListener();
+
+// This has a reference for listeners that requires to be notified, TokenOrchestrator use this for load tokens
+let resolveInflight = () => {};
```

Comment:
The following pattern gets repeated alot, maybe we could clean this up a little 🤔 

```
resolveInflight();
resolveInflight = () => {};
```

```suggestion
// This has a reference for listeners that requires to be notified, TokenOrchestrator use this for load tokens
let inflightOAuthRequestResolver = () => {};
const resolveInFlightRequest = () => {
  inflightOAuthRequestResolver();
  inflightOAuthRequestResolver = () => {};
}
```